dcap/ql/src/quote_enclave.rs

```diff
@@ -8,17 +8,179 @@
+use mc_sgx_dcap_types::RequestPolicy;
 use mc_sgx_util::ResultInto;
-use std::{ffi::CString, os::unix::ffi::OsStrExt, path::Path};
+use once_cell::sync::Lazy;
+use std::{ffi::CString, os::unix::ffi::OsStrExt, path::Path, sync::Mutex};
+
+/// A convenience type alias for a `Result` which contains an [`Error`].
+pub type Result<T> = CoreResult<T, Error>;
+
+/// Initialization of the paths for the quoting enclaves and quote provider
+/// library
+///
+/// This should only be called once during process start up utilizing
+/// [`PathInitializer::with_paths()`] or [`PathInitializer::try_default()`].
+/// If a consumer of this crate does not explicitly initialize the paths, then
+/// they will be defaulted on the first call to an SGX function that needs the
+/// paths set.
+#[derive(Debug)]
+pub struct PathInitializer;
+
+static PATH_INITIALIZER: Lazy<Mutex<Option<PathInitializer>>> = Lazy::new(|| Mutex::new(None));
+
+impl PathInitializer {
+    /// Try to initialize the paths to the default for the system
+    ///
+    /// Currently the defaults assume the default DCAP install on an Ubuntu
+    /// machine.
+    ///
+    /// # Errors
+    /// * [`Error::PathsInitialized`] if the paths have been previously
+    ///   initialized.
+    /// * [`Error::Sgx`] if any of the default paths don't exist on the system.
+    pub fn try_default() -> Result<()> {
+        Self::with_paths(
+            "/usr/lib/x86_64-linux-gnu/libsgx_qe3.signed.so.1",
+            "/usr/lib/x86_64-linux-gnu/libsgx_pce.signed.so.1",
+            None::<&Path>,
+            "/usr/lib/x86_64-linux-gnu/libsgx_id_enclave.signed.so.1",
+        )
+    }
+
+    /// Initialize the DCAP quoting library paths with provided values
+    ///
+    /// # Arguments
+    /// * `quoting_enclave` - The full file path to the quoting enclave
+    /// * `provisioning_certificate_enclave` - The full file path to the
+    ///   provisioning certificate enclave
+    /// * `quote_provider_library` - The full file path to the quote provider
+    ///   library.  When this is `None` then no quote provider library will be
+    ///   used limiting quote generation to local quote generation only.
+    /// * `id_enclave` - The full file path to the ID enclave.
+    ///
+    /// # Errors
+    /// * [`Error::PathsInitialized`] if the paths have been previously
+    ///   initialized.
+    /// * [`Error::PathStringConversion`] if one of the paths cannot be
+    ///   converted to a [`CString`]
+    /// * [`Error::Sgx`] if
+    ///     * one of the paths does not point to a file
+    ///     * one of the paths is longer than 259 (bytes)
+    ///     * one of the paths contains a null (0) byte.
+    pub fn with_paths<P1, P2, P3, P4>(
+        quoting_enclave: P1,
+        provisioning_certificate_enclave: P2,
+        quote_provider_library: Option<P3>,
+        id_enclave: P4,
+    ) -> Result<()>
+    where
+        P1: AsRef<Path>,
+        P2: AsRef<Path>,
+        P3: AsRef<Path>,
+        P4: AsRef<Path>,
```

Comment:
Optional/Equivalent: Replace some of the generic types with `impl AsRef<Path>`
```suggestion
    pub fn with_paths<P: AsRef<Path> = &Path>(
        quoting_enclave: impl AsRef<Path>,
        provisioning_certificate_enclave: impl AsRef<Path>,
        quote_provider_library: Option<P>,
        id_enclave: impl AsRef<Path>,
    ) -> Result<()>
```

I'm not 100% sure about `P: AsRef<Path> = &Path`; the idea is to allow passing `None` without naming a type

Reply:
Regarding the `P: AsRef<Path> = &Path`, https://github.com/rust-lang/rust/issues/36887
It looks like it's not meant to work on functions :(